linkerd/distribute/src/service.rs

```diff
@@ -0,0 +1,317 @@
+use super::{Distribution, WeightedKeys};
+use indexmap::IndexMap;
+use linkerd_stack::Service;
+use rand::{
+    distributions::{Distribution as _, WeightedError},
+    rngs::SmallRng,
+    SeedableRng,
+};
+use std::{
+    hash::Hash,
+    sync::Arc,
+    task::{Context, Poll},
+};
+
+/// A service that distributes requests over a set of backends.
+#[derive(Debug)]
+pub struct Distribute<K, S> {
+    backends: IndexMap<K, S>,
+    selection: Selection<K>,
+
+    /// Stores the index of the backend that has been polled to ready. The
+    /// service at this index will be used on the next invocation of
+    /// `Service::call`.
+    ready_idx: Option<usize>,
+}
+
+/// Holds per-distribution state for a [`Distribute`] service.
+#[derive(Debug)]
+enum Selection<K> {
+    Empty,
+    FirstAvailable,
+    RandomAvailable {
+        keys: Arc<WeightedKeys<K>>,
+        rng: SmallRng,
+    },
+}
+
+// === impl Distribute ===
+
+impl<K: Hash + Eq, S> Distribute<K, S> {
+    pub(crate) fn new(backends: IndexMap<K, S>, dist: Distribution<K>) -> Self {
+        Self {
+            backends,
+            selection: dist.into(),
+            ready_idx: None,
+        }
+    }
+}
+
+impl<Req, K, S> Service<Req> for Distribute<K, S>
+where
+    K: Hash + Eq,
+    S: Service<Req>,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = S::Future;
+
+    /// Acquires a ready backend.
+    ///
+    /// Note that this doesn't necessarily drive all backend services to
+    /// readiness. We expect that these inner services should be buffered or
+    /// otherwise drive themselves to readiness (i.e. via SpawnReady).
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        // If we've already chosen a ready index, then skip polling.
+        if self.ready_idx.is_some() {
+            return Poll::Ready(Ok(()));
+        }
+
+        match self.selection {
+            Selection::Empty => {
+                tracing::debug!("empty distribution will never become ready");
+            }
+
+            Selection::FirstAvailable => {
+                for (idx, svc) in self.backends.values_mut().enumerate() {
+                    if svc.poll_ready(cx)?.is_ready() {
+                        self.ready_idx = Some(idx);
+                        return Poll::Ready(Ok(()));
+                    }
```

Comment:
this is probably not worth worrying about for now, but it occurs to me that it might be possible to have a scheme where, if we've polled all the backends and none are ready, when we get woken up by one becoming ready, we could know which index became ready rather than having to poll them all in order again by wrapping the `Context`'s waker with one that knows the index and tells us which one became ready. this would avoid a second O(_n_) polling loop.

probably not worth worrying about right now, but could be an interesting optimization for the future if we care about this.

Reply:
it's not particularly easy to reset the weights... i think it's only worth yakshaving if we can identify problems with the easier approach